pkg/gadget-collection/gadgets/traceloop/gadget.go

```diff
@@ -52,138 +69,426 @@ some differences:
+		trace.Status.OperationError = fmt.Sprintf("\"collect\" operation can only be used with %q trace while %q was given", gadgetv1alpha1.TraceOutputModeStream, trace.Spec.OutputMode)
+
+		return
+	}
+
+	traceUnique.Lock()
+	if traceUnique.tracer == nil {
+		traceUnique.Unlock()
+
+		trace.Status.OperationError = "Traceloop tracer is nil"
+
+		return
+	}
+	traceUnique.Unlock()
+
+	containerID := trace.Spec.Parameters["containerID"]
+	_, ok := t.containerIDs[containerID]
+	if !ok {
+		ids := make([]string, len(t.containerIDs))
+		i := 0
+		for id := range t.containerIDs {
+			ids[i] = id
+			i++
+		}
+
+		trace.Status.OperationError = fmt.Sprintf("%q is not a valid ID for this trace, valid IDs are: %v", containerID, strings.Join(ids, ","))
+
+		return
+	}
+
+	traceUnique.Lock()
+	events, err := traceUnique.tracer.Read(containerID)
+	traceUnique.Unlock()
+	if err != nil {
+		trace.Status.OperationError = fmt.Sprintf("Failed to read perf buffer: %s", err)
+
+		return
 	}
 
-	err := t.cmd.Process.Signal(syscall.SIGINT)
+	traceName := gadgets.TraceName(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name)
+	r, err := json.Marshal(events)
 	if err != nil {
-		return fmt.Errorf("failed to send SIGINT to process: %w", err)
+		log.Warnf("Gadget %s: error marshalling event: %s", trace.Spec.Gadget, err)
+		return
 	}
+	// HACK Traceloop is really particular as it cannot use Status output because
+	// the size is limited.
+	// Also, if we send each event as a line, we will only be able to get the last
+	// 100 (or 250) events from the CLI due to this code:
+	// https://github.com/inspektor-gadget/inspektor-gadget/blob/9c7b6a126d82b54262ffdc5709d7c92480002830/pkg/gadgettracermanager/stream/stream.go#L24
+	// https://github.com/inspektor-gadget/inspektor-gadget/blob/9c7b6a126d82b54262ffdc5709d7c92480002830/pkg/gadgettracermanager/stream/stream.go#L95-L100
+	// To overcome this limitation, we just send all events as one big line.
+	// Then, the CLI receives this big line, parses it and prints each event.
+	// A proper solution would be to develop a specific "output" (neither status
+	// nor stream) for traceloop, this is let as TODO.
+	t.helpers.PublishEvent(traceName, string(r))
 
-	timeout := time.After(2 * time.Second)
+	trace.Status.State = gadgetv1alpha1.TraceStateCompleted
+}
 
-	done := make(chan struct{})
-	go func() {
-		t.cmd.Process.Wait()
-		done <- struct{}{}
-	}()
+func (t *Trace) Delete(trace *gadgetv1alpha1.Trace) {
+	if trace.Spec.OutputMode != gadgetv1alpha1.TraceOutputModeStatus {
+		trace.Status.OperationError = fmt.Sprintf("\"delete\" operation can only be used with %q trace while %q was given", gadgetv1alpha1.TraceOutputModeStatus, trace.Spec.OutputMode)
 
-	select {
-	case <-timeout:
-		t.cmd.Process.Kill()
-	case <-done:
+		return
 	}
 
-	t.started = false
+	traceUnique.Lock()
+	if traceUnique.tracer == nil {
+		traceUnique.Unlock()
+
+		trace.Status.OperationError = "Traceloop tracer is nil"
+
+		return
+	}
+	traceUnique.Unlock()
```

Comment:
This test on traceUnique.tracer does not seem to serve a purpose since your do the unlock before locking again line 406. Shouldn't you do the test just before the `traceUnique.tracer.Delete` while keeping the lock?